AMP support for AEM core components

gaetanmarmasse opened the pull request:
<!--
Before making a PR please make sure to read our contributing guidelines
https://github.com/adobe/aem-core-wcm-components/blob/master/CONTRIBUTING.md

IMPORTANT: Please base your pull request on the **development** branch! The maintainers will cherry-pick the change to
 master after it's successfully integrated and tested.

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/)
followed by the ticket number fixed by the PR. It should be underlined in the preview if done correctly.
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | N/A
| Patch: Bug Fix?          | No
| Minor: New Feature?      | Yes, AMP Support
| Major: Breaking Change?  | No
| Tests Added + Pass?      | Yes
| Documentation Provided   | Yes (code comments, README.md edited and added AMP.md)
| Any Dependency Changes?  | Added `core.wcm.components.core` dependency in `examples` module. (check 321792300b6a5d8d2f5c8e030604192d351c4cf9)
| License                  | Apache License, Version 2.0

<!-- Describe your changes below in as much detail as possible -->
**This PR adds AMP support for the AEM core components.**
Some basic technical documentation is provided in the README.md and AMP.md files and more will come in the coming days respective to each component that require mapping to an AMP component.

**Things to note / Laundry list of nice to have:**
- Bug tracked internally in order to come up with a better way to handle the uniqueness of the Navigation component in AMP and the fact that currently the ID attribute is hardcoded and tied to the show-menu component in order to allow for menu toggle. This will be updated in the coming days. [DONE]
- We may update this PR to add support for the Embed and the Experience Fragment components soon. [Postponed]
- We will likely update this PR in the coming days to fix a few minor styling issues [DONE]
- We will likely update this PR in the coming days to add a button on all pages of the Component Library site for users to be able to select the AMP version of a page. this will simply add the `.amp` sling selector to the current page the user is on. [Postponed]

richardhand commented:
Hi @gaetanmarmasse, apologies for the delay. We recently updated the component library to split it into `ui.apps` and `ui.content` (68c4071bd6f94de7), would you be able to fix the merge conflicts here due to those changes? 

I installed the PR on latest AEM master and found that the transformation seems to be working well for the different AMP Modes - with the correct html in the page for the components that have an amp alternative and the correct relative links in the header.

I wasn't able to get the injection of the `customheadlibs.amp.html` to work - they're missing in the page head and therefore the AMP components don't render - perhaps I'm missing something obvious?

http://localhost:4502/content/core-components-examples/library/carousel.amp.html

<img width="868" alt="Screenshot 2019-12-12 at 10 57 40" src="https://user-images.githubusercontent.com/25616660/70702399-4a15ba00-1cce-11ea-8077-4c025347c18a.png">

gaetanmarmasse commented:
Hi @richardhand , thanks for taking a look and letting me know. 
I have fixed the merge conflicts.
I have also moved all the local components from the examples module to their respective new home in `ui.apps` [(see 5cd50946b027b1ef2968995d43c0c4e680da0a7d)](https://github.com/adobe/aem-core-wcm-components/pull/861/commits/5cd50946b027b1ef2968995d43c0c4e680da0a7d). This was actually the issue you were having with the carousel page for instance:
1) `/content/core-components-examples/library/carousel.html` was trying to include the carousel component from `/apps/core-components-examples/components/carousel` 
2) It didn't exist because it remained in the old structure of the `examples` module, outside of `ui.apps`
3) therefore it was not rendered and produced teh following error: 

> 12.12.2019 11:07:04.743 *ERROR* [0:0:0:0:0:0:0:1 [1576174024560] GET /content/core-components-examples/library/carousel.html HTTP/1.1] org.apache.sling.servlets.get.impl.DefaultGetServlet No renderer for extension html, cannot render resource JcrNodeResource, type=core-components-examples/components/carousel, superType=null, path=/content/core-components-examples/library/carousel/jcr:content/root/responsivegrid/demo_181646866/component/carousel

<img width="300" alt="Screen Shot 2019-12-12 at 11 54 37 AM" src="https://user-images.githubusercontent.com/1090437/70740390-4b4ae500-1cd6-11ea-8db3-2ad2638935cf.png">

Please also note that I have increased the Core models version to `12.12.0` in order for the build to be successful.
Let me know if you have questions.

jckautzmann commented:
@gaetanmarmasse Thx for your PR! Before diving into the details of this PR, I’d like to discuss some broader topics.
To comply with AMP requirements, different markups need to be included in the page head:
1) the html/amp page links
2) the js links
3) the inline css
In your implementation you use different mechanisms/services to include this markup:
For 1) and 2) you use a Sling Transformer to define and include the markup.
For 3) you use a model and include it with the `page/ampheadlibs.html` HTL template.
I would prefer that we have a unique mechanism and that we replace the Sling Transformer with a dedicated model that is used in the page HTL template.
We could introduce an AMP page model (called e.g. `AmpPage`) that has the following methods:
```
getComponentCssStyles
getComponentJsLinks
getPageLinks
```
and use this model to populate the `page/ampheadlibs.html `HTL template.
I would also introduce properties (name, url) to the AMP component (under apps) to define the JS link of the AMP component instead of retrieving the content of the `customheadlibs.amp.html` file.

WDYT?

gaetanmarmasse commented:
Hi @jckautzmann , that sounds great, 
I would love to discuss these topics.

The reasoning behind retrieving the content of `customheadlibs.amp.html` is that:
1. A component can use multiple AMP libraries (i.e the search component).
2. It provides a way to add Javascript configuration objects in the head for components that might need it.

We reviewed these with @gabrielwalt as we were implementing already but yeah, I'm all for a unique mechanism to do all this so yeah, let's discuss these, :)

gaetanmarmasse commented:
For transparency, adding the outcome of the conversation regarding the technical approach for inline CSS, JS inclusion and link manipulation in the `head` raised by @jckautzmann earlier.

https://github.com/adobe/aem-core-wcm-components/pull/861/commits/4d1e8394186237f94511379f0d3555e29c41b3bc adds a model that can be directly accessed via HTL in the page component.

It simply "wraps" the existing JS library aggregation as well as it provides the proper attributes to inject in the `<link>` element.
With this approach, we are moving away from using a Transformer which made things a little less explicit from a developer standpoint.

richardhand commented on content/src/content/jcr_root/apps/core/wcm/components/accordion/v1/accordion/amp.html:
```diff
@@ -0,0 +1,37 @@
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/-->
+<amp-accordion data-sly-use.accordion="com.adobe.cq.wcm.core.components.models.Accordion"
+     class="cmp-accordion"
+     data-cmp-is="accordion"
```
`data-cmp-is` attributes can be removed in the amp renderings, they initialise the component widgets in the standard rendering.

richardhand commented on content/src/content/jcr_root/apps/core/wcm/components/accordion/v1/accordion/amp.html:
```diff
@@ -0,0 +1,37 @@
+      <section data-sly-test="${accordion.items.size > 0}"
+        data-sly-repeat.item="${accordion.items}"
+        class="cmp-accordion__item"
+        data-cmp-hook-accordion="item"
```
all `data-cmp-hook-<component>` attributes can be removed in the amp renderings, they identify sub components for widgets in the standard rendering. Same comment for the Form Text, Search, Tabs, Carousel and Image components.

richardhand commented on content/src/content/jcr_root/apps/core/wcm/components/accordion/v1/accordion/amp.html:
```diff
@@ -0,0 +1,37 @@
+        </h3>
+        <div data-sly-resource="${item.name @ decorationTagName='div'}"
+            data-cmp-hook-accordion="panel"
+            class="cmp-accordion__panel${item.name in accordion.expandedItems ? ' cmp-accordion__panel--expanded' : ' cmp-accordion__panel--hidden'}"
```
The `cmp-accordion__panel--expanded` and `cmp-accordion__panel--hidden` classes can be removed here as they don't get toggled in the amp JS.

richardhand commented on content/src/content/jcr_root/apps/core/wcm/components/sharing/v1/sharing/amp.html:
```diff
@@ -0,0 +1,26 @@
+     data-sly-use.socialMedia="com.adobe.cq.wcm.core.components.models.SocialMediaHelper"
+     data-sly-call="${template.placeholder @ isEmpty=!socialMedia.socialMediaEnabled}"></sly>
+
+<sly data-sly-test="${socialMedia.facebookEnabled && socialMedia.facebookAppId}">
```
@gaetanmarmasse - If I understand correctly, this is a new configuration. I don't see it exposed in the the `SocialMediaHelper`, perhaps I'm missing it?

gaetanmarmasse replied:
You are right, It is missing. 
Looks like it got removed when we did the rebase from `development`.
I'll update.

gaetanmarmasse replied:
Should be addressed by https://github.com/adobe/aem-core-wcm-components/pull/861/commits/554e99c40fac68ee06af4e5cddb942bcb213e44b.

richardhand commented on content/src/content/jcr_root/apps/core/wcm/components/search/v1/search/amp.html:
```diff
@@ -0,0 +1,50 @@
+  ~ limitations under the License.
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/-->
+
+<section class="cmp-search" role="search"
```
@gaetanmarmasse - I wasn't able to get the search component to display results on an AMP page when confirming an input. Following console message is displayed:
<img width="550" alt="Screenshot 2020-01-23 at 17 20 18" src="https://user-images.githubusercontent.com/25616660/73012728-373aed00-3e17-11ea-8337-6e2a71c31933.png">

gaetanmarmasse replied:
@richardhand , This is happening if you try to add the search component in the editable template without specifying the `searchTermMinimumLength="1"`.

`amp-autocomplete`, requires a JSON with actual data in it. We found it best to not modify the `searchresults` servlet provided with core components and call it with the "*" fulltext search. 
This will work only if there is an associated content policy that sets the `searchTermMinimumLength` to 1. Default is 3 which will result in an empty JSON.

richardhand commented on content/src/content/jcr_root/apps/core/wcm/components/carousel/v1/carousel/amp.html:
```diff
@@ -0,0 +1,41 @@
+    <amp-carousel data-sly-test="${carousel.items && carousel.items.size > 0}"
+      class="cmp-carousel__content"
+      type="slides"
+      width="350"
```
@gaetanmarmasse - how did you decide on this default width 350 and height 200 for the carousel? It seems a little opinionated, and can lead to strange renderings given that a carousel can have different components as items with varying content. As far as I can tell there isn't a general adaptive height solution in the `amp-carousel`.

gaetanmarmasse replied:
@richardhand agreed.
`amp-carousel` requires the `height` and `width` attributes to be set on the element. This is something we've struggled with a little on the frontend side and I believe it is a common challenge with amp around displaying images or elements without known dimensions.
We are following this recommendation: https://amp.dev/documentation/examples/style-layout/how_to_support_images_with_unknown_dimensions/
That said, I agree that this seems very opinionated and misleading. Because we are relying on CSS anyway for the slides within a carousel, the only thing I can think of is to make the `amp-carousel` attributes as followed: `height="1" width="1"`.
This would still work but may not prevent the flicker on the page if the carousel component is high / above the fold...
So basically, these sizes (350 x 200) were our best guess to make sure the element occupies the space it needs and avoid the flicker but I'm open to suggestions. ;)
WDYT?

richardhand commented on content/src/content/jcr_root/apps/core/wcm/components/tabs/v1/tabs/amp.html:
```diff
@@ -0,0 +1,53 @@
+  ~ limitations under the License.
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/-->
+  
+<div data-sly-use.tabs="com.adobe.cq.wcm.core.components.models.Tabs"
```
@gaetanmarmasse - I think for the tabs AMP component we may need to add some logic that ensures that if no active item is explicitly set by an author, the first tab should be selected. This would be in line with how the existing tabs component behaves.

gaetanmarmasse replied:
I believe this is how it works today in this component. Did you notice it was not?

richardhand replied:
Tested this again, attaching an example - I would expect the first tab to be selected if no selection was explicitly defined by an author.
![tabs-no-default-active-item](https://user-images.githubusercontent.com/25616660/73167901-0699d280-40f9-11ea-9306-2b2888cea5e4.gif)

richardhand replied:
@gaetanmarmasse - created a separate pull request for this, that ensures the first item is marked active if no active item was configured: https://github.com/Bounteous-Inc/aem-core-wcm-components/pull/55

richardhand commented on content/src/content/jcr_root/apps/core/wcm/components/form/button/v2/button/amp.html:
```diff
@@ -0,0 +1,19 @@
+  ~ limitations under the License.
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/-->
+<input data-sly-use.button="com.adobe.cq.wcm.core.components.models.form.Button"
+       type="submit"
```
@gaetanmarmasse - do you know the reason why we force the `submit` type for AMP form buttons? This could lead to unexpected behaviour if the original form button isn't configured as a submit button.

gaetanmarmasse replied:
Good catch! Should be addressed by https://github.com/adobe/aem-core-wcm-components/pull/861/commits/f89a7e605f611558dcbce56fd4f3a46856f9d40e.
There is no need for an AMP specific button markup.

richardhand commented on content/src/content/jcr_root/apps/core/wcm/components/form/options/v2/options/amp.html:
```diff
@@ -0,0 +1,39 @@
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/-->
+<sly data-sly-use.options="com.adobe.cq.wcm.core.components.models.form.Options"
```
@gaetanmarmasse - I only see the following diff from the base markup, basically the re-ordering of one attribute. Do we need this amp-specific rendering for the form options component?
<img width="1004" alt="Screenshot 2020-01-29 at 12 11 08" src="https://user-images.githubusercontent.com/25616660/73352066-9c1b9a80-4290-11ea-8238-36c4c680d356.png">

gaetanmarmasse replied:
Nope, we don't, good catch. Created https://github.com/Bounteous-Inc/aem-core-wcm-components/pull/56 that simply removes the file.

jakublamprecht commented:
Hey guys, do you have any details on release date of this AMP support? There's a project that we would possibly want to use this feature in but its highly dependant on the time it's available for use. This post (https://www.bounteous.com/insights/2019/11/15/open-source-google-amp-aem-core-components/) mentions early 2020, do you think it will be released in Q1?

jakublamprecht commented:
Hey,

We'd really like to use this feature, however we cannot wait for the official release with our development. We'd like to fork this feature and use it, but we lack information on what's been finished and what still requires some more work. Could you provide such information? Of course we would be glad to share the outcome and fixes once we're done with them.

gabrielwalt commented:
Hi @jakublamprecht, as we merged the PR, you can start to use it as a tech preview. The general architecture is finalized, but there are some known issues that will be addressed as follow-up pull requests:
* **Image:** AMP version still renders a noscript wrapper, which leads to invalid AMP, when multiple allowed widths are configured on the template in the design dialog of the component.
* **Search:** The AMP version doesn’t use the search result servlet offset and fetches all results up-front.
* **Tabs:** The unique IDs provided for the tab items need some refactoring to use the new component ID as implemented by #535 (see also #942 that is somewhat related).
* **Form Container:** Redirect page edit dialog property doesn’t seem to work with AMP, as AMP specific headers have to be set on the server.
